### ISSUE TITLE
feat: Kubeflow기반 연동 인터페이스 초기 구현

### ISSUE THREAD

**Issue by sungsoo** [traindb-project/traindb-ml#10]
### Priority-1
---

#### Deadline: 2022년 6월 28일 (화) 12시

설계해 주신 초기 연동 인터페이스 설계안(5/31)을 기반으로 아래 구현을 진행해 주세요.
구현 형상(코드 등)은 아래 github에 commit해 주세요.
https://github.com/traindb-project/traindb-ml/tree/main/interface

---

* 개요:  외부 요청(training)을 받아, 내부 pod (PyTorchJob)을 통해 ML 모델 training 및 모델 생성후, 모델을 serving하는 예제

![](https://sungsoo.github.io/images/initial-design.png)

* 구성 (그림 참조)
  * 클라이언트: Java code
  * 서버
    * TrainDB-ML:cmd: ML 모델의 (Task-1) training 요청, (Task-2) serving (ML 모델을 이용한 추론) 요청을 받는다.
    * [Task-1] POD:PyTorchJob [그림에서 POD:TFJob]: Iris 모델 생성 예제을 바탕으로 ML 모델 training 진행 후, 모델 생성 (PV:RWM)
    * [Task-2] POD:KServing: 생성된 Iris ML 모델을 이용해 inference 제공

**Comment by kihyuk-nam** [comment]
- 원격 인터페이스: TrainDB <--> TrainDB-ML/KubeFlow
  - 기능별 매핑 (SQL 기준)
    - CREATE MODELTYPE
      - 역할: (옵션: .py를 Dockerfile로 감싸서) (로컬 또는 원격의 TrainDB 전용) 레지스트리에 등록
      - 구현 방법: ... 
    - TRAIN(CREATE) MODEL
      - 구현 방법: KubeFlow Training SDK/PyTorchJob SDK 사용
        - 비교 검토: Fairing, PyTorchJob보다 약간 상위 레벨([예](https://docs.google.com/document/d/1x1JPDQfDMIbnoQRftDH1IzGU0qvHGSU4W6Jl4rJLPhI/edit#heading=h.e33ufidnl8z6)), 아니면 Pipeline 
    - SELECT
      - 구현 방법: Kserve에서 제공한 엔드포인트 주소 그대로 사용 
    - [옵션] SHOW MODELTYPEs, MODELs, MODEL_STATUS, ...
      - 역할: KubeFlow 대시보드의 Models 기능과 유사 (모델 버전, 상태, 프레임워크 종류, 엔드포인트 주소 등) 
        - 설명: https://www.kubeflow.org/docs/external-add-ons/kserve/webapp/
        - 소스: https://github.com/kserve/models-web-app/
      - 구현 방법: Kserve 이용
  - 원격 통신 방법
    - TrainDB가 있는 호스트에 k8s client(apt명 kubernetes-cli) 설치하고,
       클라이언트의 설정(~/.kube/config)에서 TrainDB-ML 주소 지정 후 직접 제어
      - cluster: 항목 아래, server: 값을 TrainDB-ML/KubeFlow 설치 서버 주소로 지정
    - 비교 검토: 별도 Knative 서비스 구현, standalone REST API 서버 구동(Fast API, Flask는 비추)

**Comment by sungsoo** [comment]
@kihyuk-nam : 아래 내용에 대해 별도의 이슈 생성함
* KubeFlow에서 지원하지 않는 ML Framework에 대한 처리 (예, (R)SPN 구현을 위한 SPFlow 라이브러리)

현재 SPFlow는 Tensorflow와 PyTorch 프레임워크를 지원: 이슈 참고

**Comment by sungsoo** [comment]
## 진행 현황 정리 요청

To @1jolee : 현재까지 진행 현황을 간략히 정리해 주세요. (**Deadline: 7/29**)
- github에 관련 내용(코드, 문서 등) commit 요청

**Comment by 1jolee** [comment]
진행 현황 정리(traindb-ml)
------------------------------
1. 2022년 7월 19일(화) : traindb-ml 회의 내용에 따라 개발 진행
2. ETRI개발서버 microk8s, kubeflow 재설치 및 정상화
3. kubernetes 및 kubeflow python SDK 설치
4. pytorchjob 및 kserve SDK 학습 및 설계에 필요한 SDK 자료 정리 중
5. POD의 구분 및 selector에 사용할 labels 정의
6. PV의 hostpath의 directory hierarchy 정의
7. traindb(java) client 개발 시작
9. traindb-ml(python SDK) program 개발 시작
    o pytorchjob SDK 먼저 진행 이후,
   ㅇkserve SDK 진행 예정 
11. 현재까지 회의 및 설계의 내용은 최상위 수준의 내용이어서, 개발을 진행하며 상세 설계를 진행하여 진행하여야 할 것으로 예상. 
12. 위와 관련 문서 및 코드는 commit할만한 상태가 아니어서, 1차 완료되면 commit 예정.
